File: app/src/main/java/com/github/goldy1992/mp3player/service/library/MediaItemTypeIds.java
package com.github.goldy1992.mp3player.service.library;

import com.github.goldy1992.mp3player.commons.MediaItemType;
import com.google.common.collect.BiMap;
import com.google.common.collect.HashBiMap;

import org.apache.commons.lang3.RandomStringUtils;

import java.util.EnumMap;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;

import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public final class MediaItemTypeIds {

    private Map<String, MediaItemType> idToMediaItemTypeMap;

    private EnumMap<MediaItemType, String> mediaItemTypeToIdMap;

    @Inject
    public MediaItemTypeIds() {
        create();
    }

    /**
     * creates the maps
     */
    private void create() {
        BiMap<MediaItemType, String> biMap = HashBiMap.create();

        // ensure unique ids
        HashSet<String> idSet = new HashSet<>();

        for (MediaItemType mediaItemType : MediaItemType.values()) {
            boolean added = false;
            String id = null;
            while (!added) {
                id = generateRootId(mediaItemType.name());
                added = idSet.add(id);
            }
            biMap.put(mediaItemType, id);
        }

        this.idToMediaItemTypeMap = new HashMap<>(biMap.inverse());

        EnumMap<MediaItemType, String> enumMap = new EnumMap<>(MediaItemType.class);
        enumMap.putAll(biMap);
        this.mediaItemTypeToIdMap = enumMap;
    }

    public String getId(MediaItemType mediaItemType) {
        return mediaItemTypeToIdMap.get(mediaItemType);
    }

    public MediaItemType getMediaItemType(String id) {
        return idToMediaItemTypeMap.get(id);
    }


    public static final String generateRootId(String prefix) {
        return prefix + RandomStringUtils.randomAlphanumeric(15);
    }
}

File: service/src/main/java/com/github/goldy1992/mp3player/service/library/MediaItemTypeIds.kt
package com.github.goldy1992.mp3player.service.library

import com.github.goldy1992.mp3player.commons.MediaItemType
import com.github.goldy1992.mp3player.service.SecureRandomUtils
import com.google.common.collect.BiMap
import com.google.common.collect.HashBiMap
import dagger.hilt.android.scopes.ServiceScoped
import java.util.EnumMap
import javax.inject.Inject

@ServiceScoped
class MediaItemTypeIds

    @Inject
    constructor() {
    private var idToMediaItemTypeMap: Map<String, MediaItemType>? = null
    private var mediaItemTypeToIdMap: EnumMap<MediaItemType, String>? = null
    /**
     * creates the maps
     */
    private fun create() {
        val biMap: BiMap<MediaItemType, String> = HashBiMap.create()
        // ensure unique ids
        val idSet = HashSet<String?>()
        for (mediaItemType in MediaItemType.values()) {
            var added = false
            var id = ""
            while (!added) {
                id = generateRootId(mediaItemType.name)
                added = idSet.add(id)
            }
            biMap[mediaItemType] = id
        }
        idToMediaItemTypeMap = HashMap(biMap.inverse())
        val enumMap = EnumMap<MediaItemType, String>(MediaItemType::class.java)
        enumMap.putAll(biMap)
        mediaItemTypeToIdMap = enumMap
    }

    fun getId(mediaItemType: MediaItemType): String {
        return mediaItemTypeToIdMap?.get(mediaItemType) ?: MediaItemType.NONE.name
    }

    fun getMediaItemType(id: String?): MediaItemType? {
        return idToMediaItemTypeMap!![id]
    }

    companion object {
        fun generateRootId(prefix: String): String {
            return prefix + SecureRandomUtils.randomAlphaNumeric(15)
        }
    }

    init {
        create()
    }
}